### PR TITLE
stylelint: correct several interfaces

### PR DESCRIPTION
**Pull request by 43081j**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://stylelint.io/developer-guide/plugins/
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

---

A few of the types were wrong.

Here's what has been done:

* Introduce postcss types to replace a few `any` (postcss is already a whitelisted dep in types-publisher)
* Add the `unix` formatter
* Use `Partial` for linter configuration (its a bad idea to have a fully optional interface)
* Replace the invalid `JSON` type (which actually refers to `JSON.parse` etc) with a new stylelint configuration interface
* Introduce a couple of missing options
* Change `ignorePath` to a string ([it is](https://stylelint.io/user-guide/node-api/#ignorepath))
* Remove `postcssResults` from linter results (stylelint's docs say this exists, however it does not exist anywhere in their source anymore. I guess it is better to not expose a non-existent thing rather than follow the probably-outdated documentation)
* Fix `createPlugin` which had the wrong interface (two params, to include the primary option)
* Add `createRuleTester` 
* Add `Plugin` interface

You can see more about these things here:
https://stylelint.io/developer-guide/plugins/

However, many of the changes were found by digging through stylelint's docs and much of their source in the spare hour I had.

### CODE REVIEW

File: types/stylelint/index.d.ts [hunk 1]
```diff
@@ -63,25 +83,74 @@ export type RuleOption = {
...
 
-    function checkAgainstRule(options: { ruleName: string; ruleSettings: any; root: any; }, callback: (warning: string) => void): void;
+    function checkAgainstRule(options: {
+        ruleName: string;
+        ruleSettings: any;
+        root: any;
+    }, callback: (warning: string) => void): void;
 }
 
+export type Plugin = (primaryOption: any, secondaryOptions: RuleOption[]) =>
```
Comment (43081j) [inline types/stylelint/index.d.ts]: this is and always was wrong, will fix tomorrow if i can (its not a `RuleOption`, what we call a `RuleOption` is in fact a validateOptions assertion object).

**Comment by 43081j** [comment]
alright this should be right now.

**This is a breaking change** so i would like to bump the version (and should) but it means it won't match with stylelint's anymore. What do people suggest? `RuleOption` has been renamed (and removed from plugins) because it was plain wrong in the previous type definitions.

**Comment by sheetalkamat** [comment]
@43081j Please refer to https://github.com/DefinitelyTyped/DefinitelyTyped#if-a-library-is-updated-to-a-new-major-version-with-breaking-changes-how-should-i-update-its-type-declaration-package to see how to update the version for breaking change.

**Comment by 43081j** [comment]
Yup I'm aware. I have been waiting for someone to review the changes first is all. 

I will bump the major version tomorrow. This will put the version out of sync with stylelint. Which is another reason I hadn't yet, as mentioned in my original post.

**Comment by 43081j** [comment]
@sheetalkamat bumped the major version. it is now out of sync with stylelint's version but seems the right thing to do.

File: types/stylelint/index.d.ts [hunk 2]
```diff
@@ -1,37 +1,56 @@
-// Type definitions for stylelint 9.4
+// Type definitions for stylelint. 10.0
```
Comment (alan-agius4) [inline types/stylelint/index.d.ts]: There is no such stylelint version.

**Review by alan-agius4 — CHANGES_REQUESTED** [review]
The major version bump is incorrect there is no such version of stylelint

**Comment by 43081j** [comment]
There doesn't need to be AFAIK @alan-agius4 

The docs specify:

> Sometimes type declaration package versions and library package versions can get out of sync.

The correct way ahead here is one of two, depending on the situation (and information i don't have on hand):

* **If the types aren't so widely used** we keep the same version aligned to stylelint its self
* **If the types are used by many** we have no choice but to bump the major version as this is indeed a breaking change to the types

You cannot break/avoid semver just to keep a version in sync with the package it provides types for. This would introduce breaking type changes into peoples builds with a minor version (or by your train of thought, a patch version, even worse). 

The reason for this unfortunately is because the original types were wrong (enough that a "clean up" is a breaking change, so, quite...).

File: types/stylelint/index.d.ts (hunk 2 above)
Comment (sheetalkamat) [reply types/stylelint/index.d.ts]: @sandersn This is correct change? The change is breaking change but library hasn't moved to new version

**Comment by 43081j** [comment]
also im still in dire need of a code review on this. i barely use stylelint and went through the sources for a few hours to get this right, so im _fairly sure_ its right but also could do with someone trying it out in a real use case.

File: types/stylelint/index.d.ts (hunk 2 above)
Comment (sandersn) [reply types/stylelint/index.d.ts]: I vote for an existing minor version, such as 9.5.

**Review by sandersn — CHANGES_REQUESTED** [review]
Well, it's a tradeoff between breaking old users &mdash; not bumping major &mdash; and confusing new users &mdash; creating types for a future version that may itself have different types.

Personally, I think the right compromise would be to bump the minor version so that things may break for old users, but hopefully in an obvious way. People who care about compatibility should have locked their dependencies anyway. And it make `@types/stylelint` easier to understand, both now and in the future when `styleline@10` is available.

**Comment by 43081j** [comment]
@sandersn WDYT about just matching the latest stylelint version? 9.10

to be honest the whole "matching up of versions" causes great headaches. I've lost count of the number of times i've seen someone ship a breaking change under a patch shamelessly because they think its more important to avoid divergence. I wish things were different.

**Comment by sandersn** [comment]
@43081j given the current situation, 9.10 seems like the best idea.

**Comment by 43081j** [comment]
@sandersn sorted.

is it worth opening an issue for discussion somewhere (if there's an appropriate place)? as this situation seems to happen relatively often

**Review by sandersn — APPROVED** [review]
The new code seems structurally reasonable but I haven't checked it against the source.

**Comment by sandersn** [comment]
@43081j I know the DT issue tracker is awash with ignored issues, but if you open an issue there and assign (or tag) me, @RyanCavanaugh and @rbuckton we are the three people who spend time thinking about Typescript + DT versioning. We can discuss there.

**Pull request merged**